### PR TITLE
[SPARK-58467][SQL] Add scalar external UDF planning support

### PR DESCRIPTION
**Pull request by haiyangsun-db**
### What changes were proposed in this pull request?

This PR adds planning support for scalar `ExternalUserDefinedFunction` expressions.

It:

- Introduces logical and physical scalar external UDF evaluation nodes.
- Extracts each UDF into a separate evaluation node. Dependent and nested UDFs become ordered, stacked nodes, so an evaluation node never contains a nested UDF call.
- Handles UDF placement around aggregates, grouping expressions, windows, and joins.
- Preserves predicate and local-limit pushdown through external UDF nodes.
- Gates planning with `spark.sql.execution.udf.unified.execution.enabled`. When disabled, external UDF expressions are rejected with a structured unsupported-feature error.
- Carries the worker specification with the UDF expression.
- Validates declared input arity and improves structured errors.
- Prevents worker specifications, payloads, and environment variables from leaking through expression, plan, error, or log rendering.
- Fails immediately for unimplemented physical execution, before starting a worker.

This PR intentionally does not implement scalar UDF execution or UDF chaining and fusing. Each UDF is currently planned as an individual evaluation node. Chaining support is tracked separately by SPARK-59049.

### Why are the changes needed?

The language-agnostic external UDF framework can represent external UDFs, but Spark did not have optimizer and planner support for scalar external UDF expressions.

The new rules provide the required placement behavior for joins, aggregates, grouping expressions, and windows while ensuring that nested UDFs do not require nested worker sessions within one evaluation node.

### Does this PR introduce _any_ user-facing change?

No. `ExternalUserDefinedFunction` currently has no public scalar entry point, and scalar execution remains unimplemented.

### How was this patch tested?

Added `PlanExternalUDFsSuite` coverage for:

- Independent, nested, deterministic, and non-deterministic UDFs.
- Aggregates, grouping expressions, windows, and joins.
- Configuration gating and unsupported placements.
- Input type and arity validation.
- Optimizer and physical planner integration.
- Safe rendering of worker specifications and environment variables.
- Fail-fast behavior for unimplemented execution.

Executed:

```bash
build/sbt -java-home /usr/lib/jvm/java-17-openjdk-amd64 \
  'sql/testOnly org.apache.spark.sql.execution.externalUDF.PlanExternalUDFsSuite' \
  sql/scalastyle \
  sql/Test/scalastyle
```

All 26 tests passed, and both Scala style checks completed without errors.

### Was this patch authored or co-authored using generative AI tooling?

Yes.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/externalUDF/PlanExternalUDFs.scala [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+          attributeMap.collectFirst {
+            case (candidate, attribute) if sameUDF(candidate, udf) => attribute
+          }
+        }
+
+        aggregate.groupingExpressions.foreach { expression =>
+          if (hasExternalUDF(expression)) {
+            val newExpression = expression.transformDown {
+              case udf: ExternalUserDefinedFunction =>
+                assert(udf.udfDeterministic,
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/externalUDF/PlanExternalUDFs.scala:169]: **Non-blocking (P2):** `PullOutNondeterministic` processes this Aggregate before `PlanExternalUDFs`, but `NondeterministicExpressionCollection` only collects `Nondeterministic` and nondeterministic `UserDefinedExpression` leaves. `ExternalUserDefinedFunction` implements neither, so a grouping UDF with `udfDeterministic = false` is left in the Aggregate and triggers the internal "should not appear in grouping expression" error before this assertion is reached. Please teach the collector to hoist nondeterministic external UDFs and add a grouping-key regression that verifies the UDF is evaluated below the Aggregate.

**Recommended change:** Teach NondeterministicExpressionCollection to hoist nondeterministic ExternalUserDefinedFunction leaves and add a focused aggregate-grouping regression test.

**Why this works:** Add an ExternalUserDefinedFunction case guarded by nondeterminism alongside the existing user-defined-expression case, allowing PullOutNondeterministic to materialize an alias in the child Project and replace the Aggregate grouping key before PlanExternalUDFs runs.

**Scope:** Catalyst nondeterministic-expression collection plus the focused scalar external-UDF planning test suite.

**Compatibility:** This changes only internal analysis of a currently failing unified external-UDF query; deterministic grouping behavior and the disabled configuration path remain unchanged.

**Risks:** The new collection case must preserve canonicalized deduplication and must not hoist deterministic external UDFs.

**Constraints:** Keep the existing PullOutNondeterministic ordering and its requirement that Aggregate grouping expressions are deterministic after rewriting.

**Success:** A nondeterministic external UDF used as both grouping key and selected grouping expression analyzes without an internal error, is evaluated in the child plan, and leaves a deterministic attribute in the Aggregate.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/externalUDF/ExtractExternalUDFFromWindow.scala [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        val windowProjectList = mutable.ArrayBuffer.empty[NamedExpression]
+        val externalUdfProjectList = window.windowExpressions.map { expression =>
+          if (containsExternalUDFOverWindowExpression(expression)) {
+            expression.transformDown {
+              case windowExpression: WindowExpression =>
+                val alias = Alias(windowExpression, s"w_${windowProjectList.size}")()
+                windowProjectList += alias
+                alias.toAttribute
+              case attribute: Attribute if !windowProjectExprIds.contains(attribute.exprId) =>
+                windowProjectList += attribute
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/externalUDF/ExtractExternalUDFFromWindow.scala]: **Nit (P3):** `Window.output` already prepends every attribute from `child.output`, so adding this ordinary child attribute to `windowProjectList` exposes the same ExprId a second time. This is reachable for `externalUDF(windowExpression, input)`. Please leave direct child attributes out of `Window.windowExpressions` and add a focused test that checks unique Window output ExprIds while preserving both UDF arguments.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/externalUDF/PlanExternalUDFs.scala [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+
+        val (udfConditions, otherConditions) = splitConjunctivePredicates(condition)
+          .partition(hasUnevaluableExternalUDF(_, join))
+        val newCondition = if (otherConditions.isEmpty) {
+          logWarning(log"The join condition:${MDC(JOIN_CONDITION, condition)} " +
+            log"of the join plan contains external UDFs only, " +
+            log"so it will be moved out and the join plan will become a cross join.")
+          None
+        } else {
+          Some(otherConditions.reduceLeft(And))
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/externalUDF/PlanExternalUDFs.scala:90]: **Nit (P3):** The suite exercises only the `otherConditions.isEmpty` path. Please add a mixed condition such as `leftKey = rightKey AND externalUDF(leftValue, rightValue)` and assert that the equality remains on the `Join` while only the external-UDF predicate moves to the post-join `Filter`; otherwise a regression here can silently turn a keyed join into a Cartesian plan.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/externalUDF/ExecuteExternalUDFExec.scala [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+ * :: Experimental ::
+ * Physical plan node that evaluates one scalar UDF in an external worker process.
+ *
+ * @param udf UDF expression evaluated by the worker session.
+ * @param resultAttr Output attribute for the UDF expression.
+ * @param child Child plan providing input rows.
+ */
+@Experimental
+case class ExecuteExternalUDFExec(
+    udf: ExternalUserDefinedFunction,
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/externalUDF/ExecuteExternalUDFExec.scala:43]: **Nit (P3):** The class Scaladoc currently says this node evaluates the UDF in a worker, but `doExecute` always throws `methodNotImplementedError`, and the PR intentionally leaves scalar execution out of scope. Please describe this as a planning placeholder with unimplemented execution until the worker path exists.

**Review by cloud-fan — COMMENTED** [review]
## Review summary

Scalar external-UDF planning is broadly structured and well covered, but a nondeterministic external UDF used as a grouping key currently fails during analysis because it is not collected by PullOutNondeterministic. I also found a duplicate-ExprId Window output path, a missing mixed-join regression test, and a physical-node Scaladoc claim that overstates the intentionally unimplemented execution behavior.

## Findings

**4 total: 0 P0, 0 P1, 1 P2, 3 P3.**

### Non-blocking (P2)

- **Hoist nondeterministic external UDF grouping keys** — `sql/core/src/main/scala/org/apache/spark/sql/execution/externalUDF/PlanExternalUDFs.scala:169` — see inline.

### Nit (P3)

- **Do not duplicate child attributes in Window output** — `sql/core/src/main/scala/org/apache/spark/sql/execution/externalUDF/ExtractExternalUDFFromWindow.scala:58` — see inline.
- **Cover mixed native and external-UDF join predicates** — `sql/core/src/main/scala/org/apache/spark/sql/execution/externalUDF/PlanExternalUDFs.scala:90` — see inline.
- **Describe scalar execution as unimplemented** — `sql/core/src/main/scala/org/apache/spark/sql/execution/externalUDF/ExecuteExternalUDFExec.scala:42` — see inline.

## Verification

- NondeterministicExpressionCollection has no matching case for ExternalUserDefinedFunction, so PullOutNondeterministic leaves the nondeterministic grouping expression in place and throws before PlanExternalUDFs runs.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/externalUDF/ExtractExternalUDFFromWindow.scala [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+          } else {
+            if (!windowProjectExprIds.contains(expression.exprId)) {
+              windowProjectList += expression
+              windowProjectExprIds += expression.exprId
+            }
+            expression.toAttribute
+          }
+        }
+        Project(
+          externalUdfProjectList,
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/externalUDF/ExtractExternalUDFFromWindow.scala]: **Blocking (P1):** This Project drops every pass-through attribute from `window.child.output`: `externalUdfProjectList` is derived only from `window.windowExpressions`, while `Window.output` is `child.output ++ windowExpressions.map(_.toAttribute)`. For a parent that selects both the input column and `externalUDF(lag(input) over (...))`, the rewritten child no longer produces the input attribute. Please include `window.child.output` in this projection in the original output order, and add an assertion that the window rewrite preserves both the input and UDF result columns.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/externalUDF/PlanExternalUDFs.scala [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+ * TODO(SPARK-55278): Revisit sharing placement logic with the Python UDF extractors after
+ * external UDF planning semantics stabilize.
+ */
+private[sql] object PlanExternalUDFs
+    extends Rule[LogicalPlan] with Logging with PredicateHelper {
+
+  override def apply(plan: LogicalPlan): LogicalPlan = plan match {
+    // A correlated subquery is rewritten as a join and revisits this rule later.
+    case subquery: Subquery if subquery.correlated => plan
+    case _ if !conf.getConf(SQLConf.UNIFIED_UDF_EXECUTION_ENABLED) =>
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/externalUDF/PlanExternalUDFs.scala]: **Blocking (P1):** This live SQLConf read can disagree with the planner retained by SessionState. The config is non-static and can be changed with `spark.conf.set`, but BaseSessionStateBuilder selects ClassicExternalUDFPlanner or UnifiedExternalUDFPlanner only once. For example, constructing SessionState with the flag enabled and then disabling it leaves the unified planner producing `MapPartitionsExternalUDF`, which this branch now rejects. Please make planner selection and optimizer gating observe the value at the same lifecycle point and cover both post-construction transitions.

**Recommended change:** Dispatch both external-UDF plan construction and optimizer gating from the current session value for each operation.

**Why this works:** Move mode selection behind a per-operation planner dispatch so both the producer and PlanExternalUDFs consume one coherent SQLConf snapshot.

**Scope:** BaseSessionStateBuilder or ExternalUDFPlanner mode dispatch, PlanExternalUDFs, and focused session-config lifecycle tests.

**Compatibility:** Preserve the existing key and fixed-value behavior while making supported runtime updates take effect consistently.

**Risks:** Planner dispatch must not create mixed classic and unified nodes within one operation.

**Constraints:** Keep disabled-mode rejection and enabled planning semantics unchanged when the value is not mutated.

**Success:** Switching the config in either direction after SessionState construction selects one coherent mode and mapInPandas no longer fails because its producer and optimizer disagree.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkOptimizer.scala [hunk 7]
```diff
@@ -147,7 +154,7 @@ class SparkOptimizer(
    * batch executing the [[ExperimentalMethods]] optimizer rules. This hook can be used to add
    * custom optimizer batches to the Spark optimizer.
    *
-   * Note that 'Extract Python UDFs' batch is an exception and ran after the batches defined here.
+   * Note that 'Extract UDFs' batch is an exception and ran after the batches defined here.
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkOptimizer.scala]: **Nit (P3):** The sentence mixes present and past tense: `is an exception and ran after`. Since this documents the current batch ordering, please use `is an exception and runs after the batches defined here`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 8]
```diff
@@ -2452,6 +2456,7 @@ object PushPredicateThroughNonJoin extends Rule[LogicalPlan] with PredicateHelpe
     case _: RebalancePartitions => true
     case _: ScriptTransformation => true
     case _: Sort => true
+    case _: ExecuteExternalUDF => true
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:2459]: **Non-blocking (P2):** The added predicate-pushdown test asserts only that the child predicate moved below `ExecuteExternalUDF`; it never checks that the predicate referencing the UDF result stays above it. A regression that drops `resultPredicate` would therefore pass while returning unfiltered rows. Please also assert an upper `Filter` whose condition is semantically equal to `resultPredicate`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/externalUDF/PlanExternalUDFs.scala [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+      case udf: ExternalUserDefinedFunction =>
+        !canEvaluate(udf, join.left) && !canEvaluate(udf, join.right)
+      case _ => false
+    }
+  }
+
+  private def extractExternalUDFFromJoinCondition(plan: LogicalPlan): LogicalPlan = {
+    plan.transformUpWithPruning(_.containsAllPatterns(EXTERNAL_UDF, JOIN)) {
+      case join @ Join(_, _, joinType, Some(condition), _)
+          if hasUnevaluableExternalUDF(condition, join) =>
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/externalUDF/PlanExternalUDFs.scala:75]: **Non-blocking (P2):** Please add the adjacent allowed-boundary case: a left outer join whose external UDF references only the left child. The test should verify that planning succeeds, `ExecuteExternalUDF` is placed below the Join on the left, and the Join condition uses its result attribute. Current cross-side tests would not catch a regression that rejects every external UDF in a non-inner ON clause.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ExternalUserDefinedFunction.scala [hunk 10]
```diff
@@ -61,6 +67,24 @@ case class ExternalUserDefinedFunction(
...
+          name = name.getOrElse(prettyName),
+          validParametersCount = Seq(types.length),
+          actualNumber = children.length)
+      case Some(types) =>
+        ExpectsInputTypes.checkInputDataTypes(children, types)
+      case None => TypeCheckSuccess
+    }
+  }
+
+  // Worker specifications and payloads can contain sensitive execution details.
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ExternalUserDefinedFunction.scala:83]: **Non-blocking (P2):** The safe-rendering contract names payloads separately from worker specifications, but the test puts its unique secret only in a worker environment variable; the payload is the public UDF name. A renderer could start appending `udf.payload` and all current secret assertions would still pass. Please use a distinct payload-only sentinel and assert it is absent from `sql`, `toString`, logged conditions, and logical and physical tree strings.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/externalUDF/PlanExternalUDFs.scala [hunk 11]
```diff
@@ -0,0 +1,266 @@
...
+    !udf.children.exists(containsExternalUDF)
+  }
+
+  private def sameUDF(
+      left: ExternalUserDefinedFunction,
+      right: ExternalUserDefinedFunction): Boolean = {
+    if (left.deterministic && right.deterministic) {
+      val normalizedPayload = Array.emptyByteArray
+      left.payload.sameElements(right.payload) &&
+        left.copy(payload = normalizedPayload).semanticEquals(
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/externalUDF/PlanExternalUDFs.scala:211]: **Non-blocking (P2):** This equality path is used specifically when mapping grouping UDFs, but the existing different-worker test nests dependent UDFs and never calls `sameUDF`. Please add a grouping case with otherwise equivalent deterministic UDFs that differ only by `workerSpec`, and assert that they produce two evaluation nodes with the correct worker/result mapping. That test should fail if canonicalization ever stops considering the worker specification.

**Review by cloud-fan — COMMENTED** [review]
## Review summary

The window rewrite drops pass-through child columns, and the unified-UDF configuration is read at incompatible lifecycle points by the session planner and optimizer. The patch also needs one Scaladoc tense fix and focused assertions for retained predicates, side-local outer joins, payload redaction, and grouping worker identity.

## Findings

**7 total: 0 P0, 2 P1, 4 P2, 1 P3.**

### Blocking (P1)

- **Preserve child columns when rewriting Window** — `sql/core/src/main/scala/org/apache/spark/sql/execution/externalUDF/ExtractExternalUDFFromWindow.scala:67` — see inline.
- **Use one lifecycle for the unified-UDF configuration** — `sql/core/src/main/scala/org/apache/spark/sql/execution/externalUDF/PlanExternalUDFs.scala:47` — see inline.

### Non-blocking (P2)

- **Assert that the UDF predicate remains above pushdown** — `sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:2459` — see inline.
- **Cover a side-local UDF in an outer join** — `sql/core/src/main/scala/org/apache/spark/sql/execution/externalUDF/PlanExternalUDFs.scala:75` — see inline.
- **Use an independent payload redaction sentinel** — `sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ExternalUserDefinedFunction.scala:83` — see inline.
- **Exercise worker identity in grouping deduplication** — `sql/core/src/main/scala/org/apache/spark/sql/execution/externalUDF/PlanExternalUDFs.scala:211` — see inline.

### Nit (P3)

- **Use present tense for the optimizer ordering** — `sql/core/src/main/scala/org/apache/spark/sql/execution/SparkOptimizer.scala:157` — see inline.

## Re-review status

**Prior AI findings: 4 addressed, 0 still present; other current findings: 7.**

New attribution: 0 newly introduced, 7 late catch, 0 previously raised, 0 unattributed.

## Remaining findings

No prior AI findings remain.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/externalUDF/PlanExternalUDFs.scala (hunk 1 above)
Comment (haiyangsun-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/externalUDF/PlanExternalUDFs.scala:169]: Fixed. `NondeterministicExpressionCollection` now hoists nondeterministic `ExternalUserDefinedFunction` expressions, and the regression test verifies that the UDF is evaluated below `Aggregate` while the grouping expressions are deterministic.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/externalUDF/ExtractExternalUDFFromWindow.scala (hunk 2 above)
Comment (haiyangsun-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/externalUDF/ExtractExternalUDFFromWindow.scala]: Fixed. Direct child attributes are no longer added to `Window.windowExpressions`. The focused regression keeps both UDF arguments and verifies that the Window output ExprIds are unique.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/externalUDF/PlanExternalUDFs.scala (hunk 3 above)
Comment (haiyangsun-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/externalUDF/PlanExternalUDFs.scala:90]: Fixed. Added a mixed join-condition regression that verifies the equality remains on the `Join` while only the external-UDF predicate moves to the upper `Filter`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/externalUDF/ExecuteExternalUDFExec.scala (hunk 4 above)
Comment (haiyangsun-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/externalUDF/ExecuteExternalUDFExec.scala:43]: Fixed. The Scaladoc now describes this as a scalar-UDF planning node and explicitly says worker execution remains unimplemented.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/externalUDF/ExtractExternalUDFFromWindow.scala (hunk 5 above)
Comment (haiyangsun-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/externalUDF/ExtractExternalUDFFromWindow.scala]: Fixed. The rewritten `Project` now prepends `window.child.output` in its original order, and the regression verifies that both the input column and UDF result are preserved.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/externalUDF/PlanExternalUDFs.scala (hunk 6 above)
Comment (haiyangsun-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/externalUDF/PlanExternalUDFs.scala]: Addressed with a different lifecycle choice. The planner is no longer retained by `SessionState`, and the unified-execution flag is now a static SQL config. Plan construction and optimizer gating therefore observe the same immutable value. Attempts to change the flag in either direction after session construction fail explicitly with `CANNOT_MODIFY_STATIC_CONFIG`, with focused coverage for both directions.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkOptimizer.scala (hunk 7 above)
Comment (haiyangsun-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SparkOptimizer.scala]: Fixed. Updated the sentence to use the present tense: “is an exception and runs after”.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 8 above)
Comment (haiyangsun-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:2459]: Fixed. The predicate-pushdown regression now verifies both sides: the child-only predicate moves below `ExecuteExternalUDF`, and the UDF-result predicate remains in the upper `Filter`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/externalUDF/PlanExternalUDFs.scala (hunk 9 above)
Comment (haiyangsun-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/externalUDF/PlanExternalUDFs.scala:75]: Fixed. Added the left-local UDF/left-outer-join case and verified that evaluation is placed below the join on the left while the join condition uses the result attribute.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ExternalUserDefinedFunction.scala (hunk 10 above)
Comment (haiyangsun-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ExternalUserDefinedFunction.scala:83]: Fixed. The rendering regression now uses separate worker-spec and payload-only sentinels and verifies that neither appears in SQL, expression strings, logged conditions, or logical and physical plan trees.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/externalUDF/PlanExternalUDFs.scala (hunk 11 above)
Comment (haiyangsun-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/externalUDF/PlanExternalUDFs.scala:211]: Fixed. Added a grouping regression with otherwise equivalent deterministic UDFs that differ only by `workerSpec`; it verifies two evaluation nodes and the correct worker/result mapping.

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
**Merge Summary:**
- merged into master https://github.com/apache/spark/commit/85494541683d66541ed1357e05b64e9094c870dc
- merged into branch-4.x https://github.com/apache/spark/commit/73f4a975da1b2c3c82dfcbb8ba01d62e438fdac9

*Posted by `merge_spark_pr.py`*
